### PR TITLE
feat: add role to credential and proof record

### PR DESCRIPTION
**Pull request by berendsliedrecht**
- Continues #1438

### CODE REVIEW

File: packages/anoncreds-rs/tests/anoncreds-flow.test.ts [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+  wallet,
+})
+
+const anoncredsCredentialFormatService = new AnonCredsCredentialFormatService()
+const anoncredsProofFormatService = new AnonCredsProofFormatService()
+
+const indyDid = 'did:indy:local:LjgpST2rjsoxYegQDRm7EL'
+
+// FIXME: Re-include in tests when NodeJS wrapper performance is improved
+describeRunInNodeVersion([18], 'AnonCreds format services using anoncreds-rs', () => {
```
Comment (TimoGlastra) [inline packages/anoncreds-rs/tests/anoncreds-flow.test.ts:77]: Is this a new test? Probably added with merge conflicts?

File: packages/anoncreds/src/updates/0.3.1-0.4/__tests__/credentialExchangeRecord.test.ts [hunk 2]
```diff
@@ -138,22 +138,50 @@ describe('0.3.1-0.4.0 | AnonCreds Migration | Credential Exchange Record', () =>
       })
     })
   })
+
+  describe('migrateRole()', () => {
```
Comment (TimoGlastra) [inline packages/anoncreds/src/updates/0.3.1-0.4/__tests__/credentialExchangeRecord.test.ts:142]: This should be moved to 0.5 migration

File: packages/anoncreds/src/updates/0.3.1-0.4/__tests__/proofExchangeRecord.test.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import { ProofExchangeRecord, ProofRepository, JsonTransformer, ProofState, ProofRole } from '../../../../../core/src'
+import { Agent } from '../../../../../core/src/agent/Agent'
+import { getAgentConfig, getAgentContext, mockFunction } from '../../../../../core/tests'
+import * as testModule from '../proofExchangeRecord'
+
+const agentConfig = getAgentConfig('AnonCreds Migration - Proof Exchange Record - 0.3.1-0.4.0')
```
Comment (TimoGlastra) [inline packages/anoncreds/src/updates/0.3.1-0.4/__tests__/proofExchangeRecord.test.ts:6]: Same, should be 0.5

File: packages/anoncreds/src/updates/0.3.1-0.4/__tests__/proofExchangeRecord.test.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    // according to: https://github.com/hyperledger/aries-rfcs/blob/main/features/0037-present-proof/README.md#states
+    genMigrateRoleTests(agent, ProofState.RequestSent, ProofRole.Verifier)
+    genMigrateRoleTests(agent, ProofState.ProposalReceived, ProofRole.Verifier)
+    genMigrateRoleTests(agent, ProofState.RequestReceived, ProofRole.Verifier)
+    genMigrateRoleTests(agent, ProofState.PresentationReceived, ProofRole.Verifier)
+
+    genMigrateRoleTests(agent, ProofState.RequestReceived, ProofRole.Prover)
+    genMigrateRoleTests(agent, ProofState.ProposalSent, ProofRole.Prover)
+    genMigrateRoleTests(agent, ProofState.PresentationSent, ProofRole.Prover)
+  })
```
Comment (TimoGlastra) [inline packages/anoncreds/src/updates/0.3.1-0.4/__tests__/proofExchangeRecord.test.ts:58]: This is not testing state done, which was still a fixme in the previous PR. I think same for credential

File: packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts [hunk 5]
```diff
@@ -36,6 +44,91 @@ export async function migrateCredentialExchangeRecordToV0_4<Agent extends BaseAg
...
+  const didCommCredentialIssuerAsReceiverRoles = [CredentialState.RequestReceived, CredentialState.ProposalReceived]
+
+  // Roles only the holder can have, as a recipient, when issuing a credential
+  const didCommCredentialHolderAsReceiverRoles = [CredentialState.OfferReceived, CredentialState.Done]
+
+  // Fetch the associated didcomm message record
+  const didCommMessageRepository = agent.dependencyManager.resolve(DidCommMessageRepository)
+  const didCommMessageRecord = await didCommMessageRepository.findSingleByQuery(agent.context, {
+    associatedRecordId: credentialRecord.id,
+  })
```
Comment (TimoGlastra) [inline packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts:85]: There can be multiple didcomm message records retuned for this. Will be good to add a full test that reaches this code with an actual database populated in 0.4 and migrating it to 0.5

File: packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts [hunk 6]
```diff
@@ -36,6 +44,91 @@ export async function migrateCredentialExchangeRecordToV0_4<Agent extends BaseAg
...
+
+  if (didCommMessageRecord) {
+    // Check whether the role of the didcomm message is `receiver`
```
Comment (TimoGlastra) [inline packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts:88]: This logic doesn't make sense. we should fetch a specific didcomm message record.

File: packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts [hunk 7]
```diff
@@ -36,6 +44,91 @@ export async function migrateCredentialExchangeRecordToV0_4<Agent extends BaseAg
...
+      // Check whether the state of the credential record is included in the allowed roles for the holder
+      if (didCommCredentialHolderAsReceiverRoles.includes(credentialRecord.state)) {
+        return CredentialRole.Holder
+      }
+    }
+
+    // Check whether the role of the didcomm message is `sender`
+    if (didCommMessageRecord.role === DidCommMessageRole.Sender) {
+      // Check whether the state of the credential record is included in the allowed roles for the issuer
+      if (didCommCredentialIssuerAsSenderRoles.includes(credentialRecord.state)) {
```
Comment (TimoGlastra) [inline packages/anoncreds/src/updates/0.3.1-0.4/credentialExchangeRecord.ts:104]: In all cases except done we can extract the role based on state, so this feels a bit redundant.
We can just check on the offer message, if role == sender, we're role = issuer else role = holder.

Might need to handle the case where there's only a proposal or only a request (if state is abandoned)

**Review by TimoGlastra — CHANGES_REQUESTED** [review]
It's not really updated for 0.5

**Comment by berendsliedrecht** [comment]
Not entirely sure why tests are failing on the `agent.config`...

https://github.com/openwallet-foundation/credo-ts/actions/runs/7958832675/job/21724520202?pr=1764#step:5:675

**Comment by berendsliedrecht** [comment]
Might need some help with setting up new snapshots :).

**Comment by TimoGlastra** [comment]
> Might need some help with setting up new snapshots :).

Did you already address this? Anything I can help with?

**Review by TimoGlastra — APPROVED** [review]

**Comment by TimoGlastra** [comment]
Should be ready now @berendsliedrecht

**Pull request merged**